app/core/utils.py

```diff
@@ -136,20 +147,75 @@ def _mode_to_result_and_type(
     return result_type, result
 
 
+connection_ttl_cache = TTLCache(maxsize=1, ttl=CONNECTION_STATUS_TTL_SECS)
+
+
+@cached(connection_ttl_cache)
+def is_connected() -> bool:
+    """Check if the system can establish a network connection to the test host. Result is cached for a short time."""
+    try:
+        socket.setdefaulttimeout(SOCKET_TIMEOUT)
+        socket.socket(socket.AF_INET, socket.SOCK_STREAM).connect((CONNECTION_TEST_HOST, CONNECTION_TEST_PORT))
+        return True
+    except socket.error:
+        return False
+
+
+def wait_with_exponential_backoff(
+    condition: Callable,
+    initial_delay: float = DEFAULT_POLLING_INITIAL_DELAY,
+    exponential_backoff: float = DEFAULT_POLLING_EXPONENTIAL_BACKOFF,
+    timeout_sec: float = DEFAULT_POLLING_TIMEOUT_SEC,
+) -> bool:
+    """
+    Wait for the condition to be True or the timeout_sec to be reached, with configurable exponential backoff.
+    Returns True if the condition is met within the timeout_sec and False otherwise.
+    """
+    start_time = time.time()
+    next_delay = initial_delay
+    target_delay = 0.0
+    while True:
+        patience_so_far = time.time() - start_time
+        if condition():
+            return True
+        if patience_so_far >= timeout_sec:
+            return False
+        target_delay = min(patience_so_far + next_delay, timeout_sec)
+        sleep_time = max(target_delay - patience_so_far, 0)
+        time.sleep(sleep_time)
+        next_delay *= exponential_backoff
+
+
+def wait_for_network_connection(
+    timeout_sec: float = DEFAULT_POLLING_TIMEOUT_SEC,
+) -> bool:
+    """
+    Wait for network connection with configurable timeout and exponential backoff.
+    Returns True if network connection is achieved within the timeout_sec and False otherwise.
+    """
+    return wait_with_exponential_backoff(is_connected, timeout_sec=timeout_sec)
+
+
 def safe_call_sdk(api_method: Callable, **kwargs):
     """
     This ensures that we correctly handle HTTP error status codes. In some cases,
     for instance, 400 error codes from the SDK are forwarded as 500 by FastAPI,
     which is not what we want.
+
+    Waits for network connection for up to one second. If no connection is found, raises a 503 error.
     """
-    try:
-        return api_method(**kwargs)
-    # except groundlight.NotFoundError as ex:
-    #     raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(ex))
-    except Exception as ex:
-        if hasattr(ex, "status"):
-            raise HTTPException(status_code=ex.status, detail=str(ex)) from ex
-        raise ex
+    has_connection = wait_for_network_connection(timeout_sec=1.0)  # Wait for connection for up to one second
```

Comment:
Doing this on every call? That seems like a bunch of extra overhead (plus at scale could get us blocked by google.)